java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java

```diff
@@ -83,13 +105,47 @@ public boolean equals(Object obj) {
     /**
      * Execution location. Currently only pre-statement, but tomorrow we might add post-statement.
      */
-    final ProgramPoint programPoint;
+    public final ProgramPoint programPoint;
     @Nullable
-    final ProgramState programState;
+    public final ProgramState programState;
+    @Nullable
+    public Node parent;
+    public final List<LearnedConstraint> learnedConstraints;
 
     Node(ProgramPoint programPoint, @Nullable ProgramState programState) {
       this.programPoint = programPoint;
       this.programState = programState;
+      learnedConstraints = new ArrayList<>();
+    }
+
+    public void setParent(@Nullable Node parent) {
+      this.parent = parent;
+      if(parent != null) {
+        programState.constraints.forEach((sv, c) -> {
+          if(parent.programState.getConstraint(sv) != c) {
+            learnedConstraints.add(new LearnedConstraint(sv, c));
+          }
+        });
+      }
+    }
+
+    public List<LearnedConstraint> learnedConstraints() {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Rename method "learnedConstraints" to prevent any misunderstanding/clash with field "learnedConstraints" defined on line 113. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1845)